src/Workspaces/Remote/Core/SolutionAssetStorage.cs

```diff
@@ -48,18 +48,20 @@ public Scope GetScope(Checksum solutionChecksum)
     /// <summary>
     /// Adds given snapshot into the storage. This snapshot will be available within the returned <see cref="Scope"/>.
     /// </summary>
-    internal ValueTask<Scope> StoreAssetsAsync(Solution solution, CancellationToken cancellationToken)
-        => StoreAssetsAsync(solution, projectId: null, cancellationToken);
+    public ValueTask<Scope> StoreAssetsAsync(Solution solution, CancellationToken cancellationToken)
+        => StoreAssetsAsync(solution.State, cancellationToken);
 
-    /// <summary>
-    /// Adds given snapshot into the storage. This snapshot will be available within the returned <see cref="Scope"/>.
-    /// </summary>
-    internal ValueTask<Scope> StoreAssetsAsync(Project project, CancellationToken cancellationToken)
-        => StoreAssetsAsync(project.Solution, project.Id, cancellationToken);
+    /// <inheritdoc cref="StoreAssetsAsync(Solution, CancellationToken)"/>
+    public ValueTask<Scope> StoreAssetsAsync(Project project, CancellationToken cancellationToken)
+        => StoreAssetsAsync(project.Solution.State, project.Id, cancellationToken);
+
+    /// <inheritdoc cref="StoreAssetsAsync(Solution, CancellationToken)"/>
+    public ValueTask<Scope> StoreAssetsAsync(SolutionState solution, CancellationToken cancellationToken)
+        => StoreAssetsAsync(solution, projectId: null, cancellationToken);
 
-    private async ValueTask<Scope> StoreAssetsAsync(Solution solution, ProjectId? projectId, CancellationToken cancellationToken)
+    /// <inheritdoc cref="StoreAssetsAsync(Solution, CancellationToken)"/>
+    public async ValueTask<Scope> StoreAssetsAsync(SolutionState solutionState, ProjectId? projectId, CancellationToken cancellationToken)
     {
-        var solutionState = solution.State;
```

Comment:
we were already transitioning from red->green here anyways, so this entire change was trivial since we don't actually need the full red snapshot at this layer at all.